src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -4,21 +4,105 @@
+    /// will throw <see cref="OperationCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resouce.</param>
+    /// <param name="targetState"></param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> </param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
+                                                     Justification = "targetState(s) parameters are mutually exclusive.")]
+    public Task WaitForResourceAsync(string resourceName, string? targetState = null, CancellationToken cancellationToken = default)
+    {
+        string[] targetStates = !string.IsNullOrEmpty(targetState) ? [targetState] : [KnownResourceStates.Running];
+        return WaitForResourceAsync(resourceName, targetStates, cancellationToken);
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach one of the specified states. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches one of the specified target states. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// If the resource doesn't reach one of the target states before <paramref name="cancellationToken"/> is signalled, this method
+    /// will throw <see cref="OperationCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="targetStates">The set of states to wait for the resource to transition to one of. See <see cref="KnownResourceStates"/> for common states.</param>
+    /// <param name="cancellationToken">A cancellation token that cancels the wait operation when signalled.</param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
+                                                     Justification = "targetState(s) parameters are mutually exclusive.")]
+    public async Task<string> WaitForResourceAsync(string resourceName, IEnumerable<string> targetStates, CancellationToken cancellationToken = default)
+    {
+        using var watchCts = CancellationTokenSource.CreateLinkedTokenSource(_applicationStopping, cancellationToken);
+        var watchToken = watchCts.Token;
+        await foreach (var resourceEvent in WatchAsync(watchToken).ConfigureAwait(false))
+        {
+            if (string.Equals(resourceName, resourceEvent.Resource.Name, StringComparisons.ResourceName)
+                && resourceEvent.Snapshot.State?.Text is { Length: > 0 } statusText
+                && targetStates.Contains(statusText, StringComparers.ResourceState))
```

Comment:
Should `targetStates` be converted to an array or some thing else, to avoid iterating over the `IEnumerable` repeatedly? I'm not sure if that is warranted here or necessarily helps a lot.

Reply:
I don't think it matters much here and it'd just be calling `Contains` anyway.